File: leetcode/easy/21-Merge-Two-Sorted-Lists.py
#!/usr/bin/python3
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
  def mergeTwoLists(self, list1: list, list2: list) -> list:
    # Base cases 
    if list1 == None:
      return list2
    if list2 == None:
      return list1
    
    # Determine which node is the smaller and recursively build the 
    # rest of the list
    if list1.val < list2.val:
      list1.next = self.mergeTwoLists(list1.next, list2)
      return list1
    else:
      list2.next = self.mergeTwoLists(list2.next, list1)
      return list2

File: neetcode150/easy/21-Merge-Two-Sorted-Lists.cpp
#include <algorithm>

/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
 
/**
 * \c Solution1
 * 
 * Recursive
 * 
 * Time  : O(n+m)
 * Space : O(n+m)
 */
class Solution1 {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if (list1 == nullptr) {
            return list2;
        }
        if (list2 == nullptr) {
            return list1;
        }
        if (list1->val < list2->val) {
            list1->next = mergeTwoLists(list1->next, list2);
            return list1;
        } else {
            list2->next = mergeTwoLists(list1, list2->next);
            return list2;
        }
    }
};

/**
 * \c Solution2
 * 
 * Iterative
 * 
 * Time  : O(n+m)
 * Space : O(1)
 */
class Solution2 {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if (list1 == nullptr) {
            return list2;
        } else if (list2 == nullptr) {
            return list1;
        }

        ListNode* curr = list1;
        if (list1->val > list2->val) {
            curr = list2;
            list2 = list2->next;
        } else {
            list1 = list1->next;
        }
        ListNode* ptr = curr;

        while (list1 && list2) {
            if (list1->val < list2->val) {
                curr->next = list1;
                list1 = list1->next;
            } else {
                curr->next = list2;
                list2 = list2->next;
            }
            curr = curr->next;
        }

        if (list1 == nullptr) {
            curr->next = list2;
        } else {
            curr->next = list1;
        }
        return ptr;
    }
};